### ISSUE TITLE
Code for new game Gin Rummy

### ISSUE THREAD

**Issue by billh0420** [datamllab/rlcard#70]
I have finished code for the card game Gin Rummy. How do I submit it if that is ok with you?

Note that the DQN training of it was very poor (essentially nothing learned). I have an option to specify an extremely simple version where the actions are essentially just discarding cards and the player scores 1 if there are no kings or queens in the hand else scores 0. This got to an average reward of 0.7 half-way through the training, but then fell down to 0.2 and stayed there.

I am not sure that I am using the training methods correctly. I just modified how Mahjong did DQN learning.

**Comment by daochenzha** [comment]
@billh0420 Hi, thanks for the efforts. We are working on a new version with slightly different interfaces. See ‘dev’ branch. You could merge the the code to dev branch if you would like to do so.

DQN is notoriously unstable. It is quite normal to observe performance drop with more training. My suggestion would be 1) try the DQN in dev branch, where the normalizer is replaced by batch normalization and could be more stable; 2) adapt the learning rate. That is, decrease the learning rate along the training process; 3) try other algorithms.

**Comment by billh0420** [comment]
I am new to GitHub. Should I read how to contribute on the web? How do I merge my code to the dev branch?

**Comment by daochenzha** [comment]
You may google how to send a pull request. Basically, you need to first fork the repo in your place. Then make changes in your local repo. Finally, send a pull request to merge to a specified branch. We will review your code once you submit the request.

**Comment by billh0420** [comment]
I started back working on the non_GUI Gin Rummy program as you suggested.

1) I fixed Gin Rummy to match the latest upgrade made to rlcard.
2) I eliminated f-strings.
3) I moved most of the testing code to tests.games.test_gin_rummy_game.py.
     The remaining testing code is not suited for unit testing. I will have to figure out what to do with    this.

Question 1: Do you want me to submit this? Without number 1) above, there are run time errors when the code in rlcard.examples are run since my code doesn't conform to the latest upgrade made.

Or, do you want me to work on the other requests that you made?

Question 2: After I would submit a request, should I stop uploading to dev until the submitted changes are pulled? In other words, does git versioning handle a subsequent push by me as a different version and it won't affect the files in the first push request?

Question 3: In the unit testing, I am testing two protected methods. Is this ok? Or, do you want me to delete those tests and test the public methods that call these protected methods?

**Comment by daochenzha** [comment]
@billh0420 Thanks for working on the non_GUI Gin Rummy!

Q1: You need to make sure that there is no run time error when sending the pull request. It would be better to send a single request after all the enhancements are finished.
Q2: Yes. Before sending the pull request, please first `git pull` since the code could be updated frequently. You can submit to the dev branch first so that we can make sure the new codes will not cause problems.
Q3: It is better only to test public methods. But I think it is also ok to test protected methods if indirectly testing it is difficult. The goal of testing is to make sure everything is right and find problems more easily in the future.

**Comment by daochenzha** [comment]
@billh0420 For the performance of Gin Rummy, in addition to tuning DQN, some engineering in reward design may also help. Currently, the reward is only in the final state, which is quite sparse. We may give some 'hints' in other states to reward some desired behaviors. An example is to give a small reward to the desired behavior (like in the scale of 0.1) and give a larger reward to the final state (like in the scale of 100). In this way, the agent could be guided by a small reward.
